File: framework/languages/dotnet/e2e/LocationMessaging/Server/Provider/ProviderHostFactory.cs
using Microsoft.Extensions.Configuration;

using LocationMessaging.Server.Provider.Configuration;
using LocationMessaging.Server.Provider.Endpoints;
using LocationMessaging.Server.Provider.Handlers;
using LocationMessaging.Server.Provider.Infrastructure;
using LocationMessaging.Shared;
using Systems.Zlink;
using Zlink.Framework.AspNetCore;
using Zlink.Framework.Contracts.Configuration;
using Zlink.Framework.Contracts.Dispatch;
using Zlink.Framework.Locations.Redis;
using Zlink.Framework.E2E.Diagnostics;

namespace LocationMessaging.Server.Provider;

internal static class ProviderHostFactory
{
    public static WebApplication Create(string[] args)
    {
        var options = ServerOptions.Parse(args, "provider");
        Directory.CreateDirectory(options.LogDir);

        var builder = WebApplication.CreateBuilder(args);
        builder.Configuration.Sources.Clear();
        builder.Configuration.AddInMemoryCollection();
        builder.Logging.ClearProviders();
        builder.Logging.AddSimpleConsole(console =>
        {
            console.SingleLine = true;
            console.TimestampFormat = "HH:mm:ss.fff ";
        });

        builder.WebHost.UseUrls(options.HttpUrl);
        var evidence = new EvidenceStore(options.Rid, options.EvidenceFile);
        builder.Services.AddSingleton(evidence);
        builder.Services.AddSingleton<BackpressureGate>();
        builder.Services.AddSingleton(new E2eMessageFlowListener(
            Path.Combine(options.LogDir, $"{options.Rid}-flow.log"),
            options.Rid,
            flow =>
            {
                if (flow.Phase is not ("error" or "dropped")) return;
                evidence.Add(
                    "dispatch-error"
                    + $"|surface={flow.Surface}"
                    + $"|kind={flow.MessageKind}"
                    + $"|reason={flow.Reason}"
                    + $"|action={flow.Action}"
                    + $"|packet={flow.PacketName ?? "<null>"}"
                    + $"|channel={flow.ChannelName ?? "<null>"}");
            }));
        if (!string.IsNullOrWhiteSpace(options.ChannelEndpoint))
            builder.Services.AddHostedService<ProfileMeshEventObserver>();

        builder.Services.AddZLinkFramework(framework =>
        {
            //  This E2E host is not started inside a memory-limited
            //  container. Supply a deterministic finite limit so the
            //  default Auto HWM contract does not depend on the host.
            var inboundDispatch = framework.ConfigureInboundDispatch();
            inboundDispatch.ProcessMemoryLimitBytes = 1UL * 1024 * 1024 * 1024;
            if (options.ApplicationHwmBytes > 0)
                inboundDispatch.ApplicationHwmBytes = options.ApplicationHwmBytes;
            // The official Redis extension registers the peer/spot/actor/route
            // stores and the owner lease store together (doc §2).
            framework.AddLocationStore(new ZLinkRedisLocationStore(redis => { redis.ConnectionString = options.RedisEndpoint
                                         ?? throw new InvalidOperationException(
                                             "Shared.RedisEndpoint is required."); redis.KeyPrefix = options.RedisKeyPrefix
                                  ?? throw new InvalidOperationException(
                                      "Shared.RedisKeyPrefix is required."); }));
            framework.ConfigureDispatch().Diagnostics
                .SetLevel(ZLinkDiagnosticsLevel.Normal);

            if (!string.IsNullOrWhiteSpace(options.ChannelEndpoint))
            {
                var profileMesh = framework.AddRouteMesh("profile")
                    .Listen(options.ChannelEndpoint)
                    .SetRoutingIdPrefix(options.Rid);
                var profile = profileMesh.Channel("profile").Server().SetWeight(options.Weight);
                var serverSocket = profileMesh.ConfigureRouterSocket();
                serverSocket.ReceiveHighWaterMark = 4;
                profile.AddRequestHandler<ProfileRequestHandler, ProfileReq, ProfileRes>("ProfileReq");
                profile.AddRequestHandler<PayloadRequestHandler, PayloadReq, PayloadRes>("PayloadReq");
                profile.AddSendHandler<ProfileCommandHandler, ProfileMsg>("ProfileMsg");
                profile.AddSendHandler<BackpressureCommandHandler, BackpressureMsg>("BackpressureMsg");
            }

            if (!string.IsNullOrWhiteSpace(options.RouteEndpoint))
            {
                var route = framework.AddRouteMesh("profile.route")
                    .Listen(options.RouteEndpoint);
                if (options.RoutePeers is { Count: > 0 })
                {
                    route.SetRoutingId(RoutingId.From(options.Rid));
                    foreach (var peer in options.RoutePeers) route.PeerConnections.Connect(peer);
                }
                else
                {
                    route.SetRoutingIdPrefix($"{options.Rid}-route");
                }

                route.AddRouteRequestHandler<RoutePingHandler, ScenarioRoutePing, ScenarioRoutePong>("ScenarioRoutePing");
            }
        });
        var app = builder.Build();
        app.MapProviderEndpoints(options);
        return app;
    }
}

File: framework/languages/node/e2e/RegistryMessaging/Server/Provider/provider-host-factory.ts
import fs from 'node:fs';
import { Module } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { ZLinkMessageFlowLogMode } from '@zlink-systems/framework';
import {
  ZLINK_FRAMEWORK_RUNTIME,
  ZLINK_ROUTE_MESH_RUNTIME_OPTIONS,
  ZLINK_ROUTE_CLIENT,
  ZLinkModule,
  zlinkFramework
} from '@zlink-systems/nestjs';
import type {
  ZLinkFrameworkRuntime,
  ZLinkRouteClient,
  ZLinkRouteMeshRuntimeOptions
} from '@zlink-systems/framework';
import { createRedisLocationStore, locationMessagingOptions } from '../../Shared/location-store';
import { PacketNames } from '../../Shared/messages';
import { validateServerOptions } from './Configuration/server-options';
import type { ServerOptions } from './Configuration/server-options';
import { REGISTRY_MESSAGING_OPTIONS, createRegistryMessagingConfigurationModule } from '../../configuration';
import { createProviderEndpoints } from './Endpoints/provider-endpoints';
import {
  EvidenceDispatchErrorObserver,
  PayloadRequestHandler,
  ProfileCommandHandler,
  ProfileRequestHandler,
  RoutePingHandler
} from './Handlers/provider-handlers';
import { EvidenceStore } from './Infrastructure/evidence-store';
import { closeHttpServer, startHttpServer } from './Support/http-server';

export async function startProviderHost(): Promise<void> {
  let stopping = false;

  const ProviderModule = createProviderModule();
  const app = await NestFactory.createApplicationContext(ProviderModule, { logger: false, abortOnError: false });
  const options = app.get(REGISTRY_MESSAGING_OPTIONS, { strict: false }) as ServerOptions;
  const evidence = app.get(EvidenceStore, { strict: false });
  const route = app.get(ZLINK_ROUTE_CLIENT, { strict: false }) as ZLinkRouteClient;
  const runtimeOptions = app.get(ZLINK_ROUTE_MESH_RUNTIME_OPTIONS, { strict: false }) as ZLinkRouteMeshRuntimeOptions;
  const frameworkRuntime = app.get(ZLINK_FRAMEWORK_RUNTIME, { strict: false }) as ZLinkFrameworkRuntime;
  const server = await startHttpServer(
    options.httpUrl,
    createProviderEndpoints(evidence, route, route, runtimeOptions, frameworkRuntime, () => { stopping = true; })
  );

  while (!stopping) {
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  await closeHttpServer(server);
  await app.close();
}

function createProviderModule(): Function {
  class ProviderModule {}
  const configuration = createRegistryMessagingConfigurationModule(validateServerOptions);

  Module({
    imports: [
      configuration,
      ZLinkModule.forRootFactory({
        imports: [configuration], inject: [REGISTRY_MESSAGING_OPTIONS],
        useFactory: (value: unknown) => {
          const options = value as ServerOptions;
          fs.mkdirSync(options.logDir, { recursive: true });
          const builder = zlinkFramework();
          builder
            .configureDispatch()
              .setMessageFlowObserver(EvidenceDispatchErrorObserver)
              .messageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
              .traceLogFile(`${options.logDir}/${options.rid}-flow.log`)
              .traceLabel(options.rid);

          if (options.redisEndpoint !== undefined && options.redisKeyPrefix !== undefined) {
            builder.addLocationStore(createRedisLocationStore({
              redisEndpoint: options.redisEndpoint,
              redisKeyPrefix: options.redisKeyPrefix
            }));
            locationMessagingOptions(builder.configureLocations());
          }
          if (options.channelEndpoint !== undefined) {
            const profile = builder.addRouteMesh('profile')
              .listen(options.channelEndpoint)
              .routingId(options.rid);
            profile.peerConnections();
            profile.channel('profile').server()
              .setWeight(options.weight)
              .addRequestHandler(PacketNames.profileReq, ProfileRequestHandler)
              .addRequestHandler(PacketNames.payloadReq, PayloadRequestHandler)
              .addSendHandler(PacketNames.profileMsg, ProfileCommandHandler);
          }
          if (options.manualClientEndpoint !== undefined) {
            builder.addRouteMesh('profile.manual')
              .peerConnections().connect(options.manualClientEndpoint);
          }
          if (options.routeEndpoint !== undefined) {
            const routeMesh = builder.addRouteMesh('profile.route')
              .listen(options.routeEndpoint)
              .routingId(options.rid)
              .addRequestHandler(PacketNames.scenarioRouteReq, RoutePingHandler);
            for (const endpoint of options.routePeers) routeMesh.peerConnections().connect(endpoint);
          }
          return builder.build();
        }
      })
    ],
    providers: [
      { provide: EvidenceStore, inject: [REGISTRY_MESSAGING_OPTIONS], useFactory: (value: unknown) => {
        const options = value as ServerOptions; return new EvidenceStore(options.rid, options.evidenceFile);
      } },
      EvidenceDispatchErrorObserver,
      PayloadRequestHandler,
      ProfileCommandHandler,
      ProfileRequestHandler,
      RoutePingHandler
    ]
  })(ProviderModule);
  return ProviderModule;
}
